### PR TITLE
Add vectored positioned I/O on Unix

### PR DESCRIPTION
**Pull request by a1phyr**
Add methods for vectored I/O with an offset on `File` for `unix` under `#![feature(unix_file_vectored_at)]`.

The new methods are wrappers around `preadv` and `pwritev`.

Tracking issue: #89517

### CODE REVIEW

**Comment by a1phyr** [comment]
Well, I copied `read_vectored` for this.

Anyway, `readv` takes a `const struct iovec*`, so this is the good type. However, according to Rust rules we cannot modify something behind a shared reference, so I think we should get a `*mut` pointer and cast it to `*const`.

**Comment by a1phyr** [comment]
Rebased

**Comment by Dylan-DPC** [comment]
@joshtriplett this is ready for review

**Comment by workingjubilee** [comment]
@a1phyr Sorry for this slipping through the cracks! If you can fix the latest merge conflicts I can get this approved.

r? @workingjubilee

**Comment by a1phyr** [comment]
Rebased

@rustbot label -S-waiting-on-author +S-waiting-on-review

**Comment by workingjubilee** [comment]
Aa, eeto... bleh!

**Comment by matthiaskrgr** [comment]
@bors r- rollup=iffy
failed in https://github.com/rust-lang/rust/pull/108559
`CANNOT LINK EXECUTABLE: could not load library "libstd-ef8e480f4327ce79.so" needed by "/data/local/tmp/work/test0/a"; caused by cannot locate symbol "preadv" referenced by "libstd-ef8e480f4327ce79.so"...`

https://github.com/rust-lang-ci/rust/actions/runs/4292167740/jobs/7478270814#step:26:16943

**Comment by workingjubilee** [comment]
@a1phyr Do you have any insight into the linkage issues?

**Comment by a1phyr** [comment]
Wild guess: `preadv` was added to bionic in 2015 (cf https://android.googlesource.com/platform/bionic/+/6f4594d). The tests may run with on older version.

To fix this, we could either disable `preadv` on android or use weak linkage.

**Comment by workingjubilee** [comment]
Our Android minimum is increasing to Android KitKat. That's lower than Marshmallow, the Oct 2015 release. I need more coffee before I remember the details of how weak symbol linkage works, but what we do here has to work on KitKat and Lollipop, so go with whichever feels right there.

File: library/std/src/sys/unix/fd.rs [hunk 1]
```diff
@@ -197,6 +281,90 @@ impl FileDesc {
...
+                cmp::min(bufs.len(), max_iov()) as libc::c_int,
+                offset as _,
+            )
+        })?;
+        Ok(ret as usize)
+    }
+
+    #[cfg(all(target_os = "android", target_pointer_width = "32"))]
+    pub fn write_vectored_at(&self, bufs: &[IoSlice<'_>], offset: u64) -> io::Result<usize> {
+        super::weak::weak!(fn pwritev64(libc::c_int, *const libc::iovec, libc::c_int, off64_t) -> isize);
```
Comment (workingjubilee) [inline library/std/src/sys/unix/fd.rs:350]: I, too, feel super weak when thinking about these sorts of things.

**Review by workingjubilee — CHANGES_REQUESTED** [review]
Bad news, I looked it up and it seems this was only introduced relatively late in the game by macOS (and iOS). It seems even https://github.com/rust-lang/rust/pull/104385 won't raise the minimum supported versions high enough to guarantee pwritev on those platforms.

**Review by workingjubilee — APPROVED** [review]
Huzzah, let's see if this goes or if it tells us to do more.

@bors r+

**Comment by workingjubilee** [comment]
Not again?
@bors r+

**Comment by rylev** [comment]
Small regression in Diesel which has started to show some bimodality. I think it's safe to call this triaged.

@rustbot label +perf-regression-triaged

**Pull request merged**